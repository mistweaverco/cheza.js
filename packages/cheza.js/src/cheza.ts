import type { ChezaDataStore, ChezaOpts } from '../types'
import { canPlayType } from './canPlayType'
import { ChezaCanPlayTypes } from '../enums'

class Cheza {
  private readonly _dataStore: ChezaDataStore
  constructor (videoElement: HTMLVideoElement, opts: ChezaOpts) {
    this._dataStore = {
      videoElement
    }
    console.log('Cheza constructor')
  }

  public static canPlayType = canPlayType
  public static CanPlayTypes = ChezaCanPlayTypes
};

export { Cheza }
