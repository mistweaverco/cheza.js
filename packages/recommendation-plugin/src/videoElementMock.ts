const parentElement = document.createElement('div')
const videoElementMock = document.createElement('video')
videoElementMock.removeAttribute('controls')

Object.defineProperty(videoElementMock, 'paused', {
  writable: true,
  value: false
})

Object.defineProperty(videoElementMock, 'duration', {
  writable: true,
  value: 100
})

Object.defineProperty(videoElementMock, 'play', {
  writable: true,
  value: () => {
  }
})

Object.defineProperty(videoElementMock, 'pause', {
  writable: true,
  value: function () {
    this.paused = true
  }
})

parentElement.appendChild(videoElementMock)

export { videoElementMock }
