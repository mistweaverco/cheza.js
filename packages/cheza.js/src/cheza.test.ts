import { Cheza } from './cheza'
import type { ChezaPlugin } from '../types'
import { videoElementMock } from './videoElementMock'

const ExampleChezaPlugin: typeof ChezaPlugin = class {
  public static Name = 'ExampleChezaPlugin'
  public static Version = '0.0.1'

  constructor () {
    console.log('ExampleChezaPlugin.constructor()')
    return this
  }

  add = (dataStore: any, opts?: any): void => {
    console.log('ExampleChezaPlugin.add()', dataStore, opts)
  }

  remove = (dataStore: any, opts?: any): void => {
    console.log('ExampleChezaPlugin.remove()', dataStore, opts)
  }
}

const cheza = new Cheza(videoElementMock)

describe('Cheza', () => {
  it('should match the dataStore snapshot', () => {
    expect(cheza.getDataStore()).toMatchSnapshot()
  })

  it('should register a plugin', () => {
    const res = Cheza.registerPlugin(ExampleChezaPlugin)
    expect(res).toBe(true)
  })

  it('should fail to register a already registered plugin', () => {
    const res = Cheza.registerPlugin(ExampleChezaPlugin)
    expect(res).toBe(false)
  })

  it('should unregister a already registered plugin', () => {
    const res = Cheza.unregisterPlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(true)
  })

  it('should fail to unregister a already unregistered plugin', () => {
    const res = Cheza.unregisterPlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(false)
  })

  it('should fail to add a not registered plugin', () => {
    const res = cheza.addPlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(false)
  })

  it('should fail to remove a not registered plugin', () => {
    const res = cheza.removePlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(false)
  })

  it('should succeed to add a previously registered plugin', () => {
    Cheza.registerPlugin(ExampleChezaPlugin)
    const res = cheza.addPlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(true)
  })

  it('should succeed to remove a previously registered and added plugin', () => {
    Cheza.registerPlugin(ExampleChezaPlugin)
    const res = cheza.removePlugin(ExampleChezaPlugin.Name)
    expect(res).toBe(true)
  })
})
