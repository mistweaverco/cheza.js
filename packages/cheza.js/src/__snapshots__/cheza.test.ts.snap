// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cheza should match the dataStore snapshot 1`] = `
{
  "activePlugins": Map {},
  "controlsBottom": <div
    class="controls-bottom"
  >
    <button
      aria-label="Play"
      class="button button-play"
      title="Play"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-play"
        />
      </svg>
    </button>
    <button
      aria-label="Pause"
      class="button button-pause hidden"
      title="Pause"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-pause"
        />
      </svg>
    </button>
    <button
      aria-label="Mute"
      class="button button-volume"
      title="Mute"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-volume"
        />
      </svg>
    </button>
    <button
      aria-label="Unmute"
      class="button button-muted hidden"
      title="Unmute"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-muted"
        />
      </svg>
    </button>
    <button
      aria-label="Enter Fullscreen"
      class="button button-enter-fullscreen"
      title="Enter Fullscreen"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-enter-fullscreen"
        />
      </svg>
    </button>
    <button
      aria-label="Exit Fullscreen"
      class="button button-exit-fullscreen hidden"
      title="Exit Fullscreen"
      type="button"
    >
      <svg
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="#cheza-exit-fullscreen"
        />
      </svg>
    </button>
  </div>,
  "controlsTop": <div
    class="controls-top"
  />,
  "loadingSpinner": <div
    class="loading-spinner"
  >
    <div
      class="animation"
    />
  </div>,
  "progress": <div
    class="progress"
  />,
  "progressContainer": <div
    class="progress-container"
  >
    <div
      class="progress-inline"
    >
      <div
        class="progress"
      />
    </div>
  </div>,
  "rootElement": <div
    class="cheza"
  >
    <div
      class="video-container"
    >
      <video />
    </div>
    <div
      class="loading-spinner"
    >
      <div
        class="animation"
      />
    </div>
    <div
      class="controls-top"
    />
    <div
      class="progress-container"
    >
      <div
        class="progress-inline"
      >
        <div
          class="progress"
        />
      </div>
    </div>
    <div
      class="controls-bottom"
    >
      <button
        aria-label="Play"
        class="button button-play"
        title="Play"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-play"
          />
        </svg>
      </button>
      <button
        aria-label="Pause"
        class="button button-pause hidden"
        title="Pause"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-pause"
          />
        </svg>
      </button>
      <button
        aria-label="Mute"
        class="button button-volume"
        title="Mute"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-volume"
          />
        </svg>
      </button>
      <button
        aria-label="Unmute"
        class="button button-muted hidden"
        title="Unmute"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-muted"
          />
        </svg>
      </button>
      <button
        aria-label="Enter Fullscreen"
        class="button button-enter-fullscreen"
        title="Enter Fullscreen"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-enter-fullscreen"
          />
        </svg>
      </button>
      <button
        aria-label="Exit Fullscreen"
        class="button button-exit-fullscreen hidden"
        title="Exit Fullscreen"
        type="button"
      >
        <svg
          focusable="false"
          role="presentation"
        >
          <use
            xlink:href="#cheza-exit-fullscreen"
          />
        </svg>
      </button>
    </div>
  </div>,
  "videoContainer": <div
    class="video-container"
  >
    <video />
  </div>,
  "videoElement": <video />,
}
`;
